src/v/storage/disk_log_impl.cc

```diff
@@ -1697,4 +1697,74 @@ log make_disk_backed_log(
     return log(ptr);
 }
 
+ss::future<reclaim_size_limits>
+disk_log_impl::estimate_reclaim_size(compaction_config cfg) {
+    if (!config().is_collectable()) {
+        co_return reclaim_size_limits();
+    }
+
+    cfg = apply_overrides(cfg);
+
+    auto max_offset = retention_offset(cfg);
+    if (!max_offset.has_value()) {
+        co_return reclaim_size_limits();
+    }
+
+    /*
+     * evicting data based on the retention policy is a coordinated effort
+     * between disk_log_impl housekeeping and the raft/eviction_stm. it works
+     * roughly as follows:
+     *
+     * 1. log computes ideal target offset for reclaim: max_offset computed by
+     *    the call to retention_offset(cfg).
+     *
+     * 2. log notifies the eviction stm which computes: min(max_offset, clamp)
+     *    where clamp is any additional restrictions, such as what has been
+     *    uploaded to cloud storage.
+     *
+     * 3. raft attempts to prefix truncate the log at the clamped offset after
+     *    it has prepared and taken any necessary snapshots.
+     *
+     * Because this effort is split between the log, raft, and eviction_stm we
+     * end up having to duplicate some of the logic here, such as handling the
+     * max collectible offset from stm. A future refactoring may consider moving
+     * more of the retention controls into a higher level location.
+     */
+    const auto max_collectible = stm_manager()->max_collectible_offset();
+    const auto retention_offset = std::min(max_offset.value(), max_collectible);
+
+    /*
+     * truncate_prefix() dry run
+     */
+    fragmented_vector<segment_set::type> retention_segments;
+    fragmented_vector<segment_set::type> available_segments;
+    for (auto& seg : _segs) {
+        if (seg->offsets().dirty_offset < retention_offset) {
+            retention_segments.push_back(seg);
+        } else if (seg->offsets().dirty_offset < max_collectible) {
```

Comment:
Should the condition here be: `seg->offsets().dirty_offset <= max_collectible`? Same for the if above.

Reply:
@VladLazar i don't think so, but i'm interested if you think otherwise. in `remove_prefix_full_segments` the predicate in `do_until(p, remove_front_segment)` is `p = _segs.front()->offsets().dirty_offset >= cfg.start_offset`, so the negation of that would be strictly `<`. wdyt?

Reply:
`log_eviction_stm` determines the offset at which the snapshot will occur [here](https://github.com/redpanda-data/redpanda/blob/8a9d43faa4b0d998eb6cd79521090b1733eca0e1/src/v/raft/log_eviction_stm.cc#L73-L80). Let's assume that's offset `X`. Then, raft requests truncates the log at `X + 1` [here](https://github.com/redpanda-data/redpanda/blob/8a9d43faa4b0d998eb6cd79521090b1733eca0e1/src/v/raft/consensus.cc#L2217-L2218). Finally, all segments with `dirty_offset < X + 1` are removed [here](https://github.com/redpanda-data/redpanda/blob/8a9d43faa4b0d998eb6cd79521090b1733eca0e1/src/v/storage/disk_log_impl.cc#L1265-L1277).

`retention_offset` in your code is offset `X` (what `log_eviction_stm` comes up with, so the comparison on line 1742 should use `<=`  (i think).

Reply:
You're right. I missed `model::next_offset(last_included_index)` when I was tracing the code yesterday. Thanks!